File: packages/wallet_core/ios/Classes/WalletManager.swift
//
//  WalletManager.swift
//  wallet_core
//
//  Created by niBVL on 31/10/2021.
//

import Foundation
import WalletCore

class WalletManager {
    
    var callData = [Any]()
    var resultData: [String: Any] = ["code": 200]
    
    func genNewWallet() {
        let strength = callData[0] as! Int32
        let passphrase = callData[1] as! String
        let wallet = HDWallet(strength: strength, passphrase: passphrase)
        createAddress(wallet)
    }
    
    func importNewWalletByMnemonic()  {
        let mnemonic = callData[0] as! String
        if HDWallet.isValid(mnemonic: mnemonic) {
            let wallet = HDWallet.init(mnemonic: mnemonic, passphrase: "")
            createAddress(wallet)
        }
        else {
            updateResultFailure(400)
        }
        
        
    }
    
    func importNewWalletByPrivateKey()  {
        let key = callData[0] as! String
        let keyData = Data.init(hexString: key)
        if let safeKeyData = keyData {
            let isValid = PrivateKey.isValid(data: safeKeyData, curve: CoinType.smartChain.curve)
            if isValid {
                let privateKey = PrivateKey(data: safeKeyData)
                if let safePrivateKey = privateKey {
                    let address = CoinType.smartChain.deriveAddress(privateKey: safePrivateKey)
                    let walletModel = WalletModel(mnemonic: "", privateKey: key, address: address )
                    updateResultSuccess(walletModel.toDic())
                    return
                }
                
            }
        }
        updateResultFailure(400)
    }
    
    
    private func createAddress(_ wallet: HDWallet) {
        let privateKey = wallet.getKey(coin: CoinType.smartChain, derivationPath: CoinType.smartChain.derivationPath())
        let address = CoinType.smartChain.deriveAddress(privateKey: privateKey)
        let walletModel = WalletModel(mnemonic: wallet.mnemonic, privateKey: privateKey.data.hexString, address: address)
        updateResultSuccess(walletModel.toDic())
        
    }
    
    private func updateResultSuccess(_ body: [String:Any]) {
        resultData.updateValue(body, forKey: "body")
        resultData.updateValue(200, forKey: "code")
    }
    
    private func updateResultFailure(_ code: Int) {
        resultData.updateValue("", forKey: "body")
        resultData.updateValue(code, forKey: "code")
    }

}

struct WalletModel {
    let mnemonic: String
    let privateKey: String
    let address: String
        
    func toDic() -> [String: String]{
        return ["mnemonic": mnemonic, "privateKey": privateKey, "address": address]
    }
    
}

File: packages/wallet_core/android/src/main/kotlin/com/packages/wallet_core/wallet_core/WalletManager.kt
package com.packages.wallet_core.wallet_core
import kotlin.experimental.and

import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import  wallet.core.jni.PrivateKey
import java.security.InvalidParameterException

class WalletManager {
    init {
        System.loadLibrary("TrustWalletCore")
    }

    var callData = listOf<Any>()
    var resultData = hashMapOf<String, Any>()

    fun genNewWallet(){
        val strength = callData[0] as Int
        val passphrase = callData[1] as String
        val wallet = HDWallet(strength, passphrase)
        createAddress(wallet)

    }

    fun importNewWalletByMnemonic() {
        val mnemonic = callData[0] as String
        try {
            val wallet = HDWallet(mnemonic, "")
            createAddress(wallet)
        }
        catch (exp: InvalidParameterException) {
            updateResultFailure(400)
        }
    }

    fun importNewWalletByPrivateKey() {
        val key = callData[0] as String
        if (key.contains("0x")) {
            key.drop(2)
        }
        val dataKey = key.hexStringToByteArray()
        val isValidKey = PrivateKey.isValid(key.hexStringToByteArray(), CoinType.SMARTCHAIN.curve())
        if (isValidKey) {
            val privateKey = PrivateKey(key.hexStringToByteArray())
            val address = CoinType.SMARTCHAIN.deriveAddress(privateKey)
            val walletModel = WalletModel("", key,address)
            updateResultSuccess(walletModel.toMap())
        }
        else {
            updateResultFailure(400)
        }
    }

    private  fun createAddress(wallet: HDWallet) {
        val privateKey =  wallet.getKey(CoinType.SMARTCHAIN, CoinType.SMARTCHAIN.derivationPath())
        val address = CoinType.SMARTCHAIN.deriveAddress(privateKey)
        val walletModel =  WalletModel(wallet.mnemonic(), privateKey.data().toHexString(false), address)
        updateResultSuccess(walletModel.toMap())

    }

    private fun updateResultSuccess(body: HashMap<String, kotlin.Any>) {
        resultData["body"] = body
        resultData["code"] = 200
    }
    private  fun updateResultFailure(code: Int) {
        resultData["body"] = ""
        resultData["code"] = code
    }

    private fun String.hexStringToByteArray() : ByteArray {
        val hexChars = "0123456789ABCDEF"
        val result = ByteArray(this.length / 2)
        for (i in 0 until length step 2) {
            val firstIndex = hexChars.indexOf(this[i].toUpperCase());
            val secondIndex = hexChars.indexOf(this[i + 1].toUpperCase());
            val octet = firstIndex.shl(4).or(secondIndex)
            result[i.shr(1)] = octet.toByte()
        }
        return result
    }

    private fun ByteArray.toHexString(withPrefix: Boolean = true): String {
        val stringBuilder = StringBuilder()
        if(withPrefix) {
            stringBuilder.append("0x")
        }
        for (element in this) {
            stringBuilder.append(String.format("%02x", element and 0xFF.toByte()))
        }
        return stringBuilder.toString()
    }

}

data class WalletModel(val mnemonic: String, val privateKey: String, val address: String) {
    fun toMap():HashMap<String, Any> {
        return hashMapOf<String, Any>("mnemonic" to mnemonic, "privateKey" to privateKey, "address" to address)
    }
}


